src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// analyses whose soundness would be affected (e.g. "data flows out"), we track "unassignments" to keep
     /// the analysis sound.
     /// </summary>
-    internal abstract partial class AbstractFlowPass<TLocalState> : PreciseAbstractFlowPass<TLocalState>
-        where TLocalState : PreciseAbstractFlowPass<TLocalState>.AbstractLocalState
+    internal abstract partial class AbstractFlowPass<TLocalState> : BoundTreeVisitor
+        where TLocalState : AbstractFlowPass<TLocalState>.ILocalState
     {
-        private readonly bool _trackUnassignments; // for the data flows out walker, we track unassignments as well as assignments
+        protected int _recursionDepth;
 
-        protected AbstractFlowPass(
-            CSharpCompilation compilation,
-            Symbol member,
-            BoundNode node,
-            bool trackUnassignments = false)
-            : base(compilation, member, node)
+        /// <summary>
+        /// The compilation in which the analysis is taking place.  This is needed to determine which
+        /// conditional methods will be compiled and which will be omitted.
+        /// </summary>
+        protected readonly CSharpCompilation compilation;
+
+        /// <summary>
+        /// The method whose body is being analyzed, or the field whose initializer is being analyzed.
+        /// It is used for
+        /// references to method parameters. Thus, 'member' should not be used directly, but
+        /// 'MethodParameters', 'MethodThisParameter' and 'AnalyzeOutParameters(...)' should be used
+        /// instead.
+        /// </summary>
+        protected readonly Symbol _member;
+
+        /// <summary>
+        /// The bound node of the method or initializer being analyzed.
+        /// </summary>
+        protected readonly BoundNode methodMainNode;
+
+        /// <summary>
+        /// The flow analysis state at each label, computed by merging the state from branches to
+        /// that label with the state when we fall into the label.  Entries are created when the
+        /// label is encountered.  One case deserves special attention: when the destination of the
+        /// branch is a label earlier in the code, it is possible (though rarely occurs in practice)
+        /// that we are changing the state at a label that we've already analyzed. In that case we
+        /// run another pass of the analysis to allow those changes to propagate. This repeats until
+        /// no further changes to the state of these labels occurs.  This can result in quadratic
+        /// performance in unlikely but possible code such as this: "int x; if (cond) goto l1; x =
+        /// 3; l5: print x; l4: goto l5; l3: goto l4; l2: goto l3; l1: goto l2;"
```

Comment:
interesting.  is the algorithm though always guaranteed to terminate?

Reply:
Ok.  So it's critical that the lattice forms a dag then.  And that any operation on any state never moves the state back toward the starting state (or even keep a state unchanged).  It's unclear how the latter is necessarily ensured.  Note: this is not a complaint, or suggestion that anything be changed.  It's more about my being able to convince myself that hte implementation behaves properly and is certain to terminate.

Do we have anything in place to *ensure* somehow that a state always moves toward the top of the lattice and never does anything like just stay in the same place?  For example, in a move step, do we assert that the new state is not equal to the state it was previously in?

Reply:
If the functions are monotonic and we stop iterating when we reach a fixed point, the proof is pretty simple. One thing I can add is an assert in definite assignment that we never call SetSlotUnassigned unless NonMonotonicTransfer is true. That should suffice for verifying we can never move up the lattice.

Reply:
Scratch that: the same function is called for declaring a fresh variable with an unassigned state. Maybe worth adding something in the future.

Reply:
@CyrusNajmabadi I'm still learning, but my understanding is that we're guaranteed to terminate because: 
1. when updating our knowledge of a point in the program, we only use `Join` (so we make progress in a single direction, regardless of whether the transfer function is monotonic or not) 
2. `Join` is capped by an upper-bound (so we can't go on forever).

Reply:
@jcouv Correct.

Reply:
> Ok.  So it's critical that the lattice forms a dag then?

Every lattice is a DAG. [https://en.wikipedia.org/wiki/Lattice_(order](https://en.wikipedia.org/wiki/Lattice_(order))

> And that any operation on any state never moves the state back toward the starting state (or even keep a state unchanged).

No, that is not a requirement.  For nullable analysis, a block of code can indeed assign a non-null value to a variable (down the lattice), and it can also assign a null value to a variable (up the lattice).  Still, it is the case that the state at any given label in the code can only move up the lattice from one iteration of analysis to the next due to the properties of the lattice and its Meet and Join operations. As such the analysis can only go a finite number of iterations before it reaches a fixed point, as it would reach the absolute top of the lattice after a finite number of steps, after which it cannot change.

---
In reply to: [236840657](https://github.com/dotnet/roslyn/pull/31389#discussion_r236840657) [](ancestors = 236840657)